metricbeat/module/nats/connection/data.go

```diff
@@ -0,0 +1,119 @@
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package connection
+
+import (
+	"encoding/json"
+
+	"github.com/pkg/errors"
+
+	s "github.com/elastic/beats/v7/libbeat/common/schema"
+	c "github.com/elastic/beats/v7/libbeat/common/schema/mapstriface"
+	"github.com/elastic/beats/v7/metricbeat/mb"
+	"github.com/elastic/beats/v7/metricbeat/module/nats/util"
+)
+
+var (
+	moduleSchema = s.Schema{
+		"server": s.Object{
+			"id":   c.Str("server_id"),
+			"time": c.Str("now"),
+		},
+	}
+	connectionsSchema = s.Schema{
+		"name":          c.Str("name"),
+		"subscriptions": c.Int("subscriptions"),
+		"in": s.Object{
+			"messages": c.Int("in_msgs"),
+			"bytes":    c.Int("in_bytes"),
+		},
+		"out": s.Object{
+			"messages": c.Int("out_msgs"),
+			"bytes":    c.Int("out_bytes"),
+		},
+		"pending_bytes": c.Int("pending_bytes"),
+		"uptime":        c.Str("uptime"),
+		"idle_time":     c.Str("idle"),
+	}
+)
+
+// Connections stores connections related information
+type Connections struct {
+	Now         string                   `json:"now"`
+	ServerID    string                   `json:"server_id"`
+	Connections []map[string]interface{} `json:"connections,omitempty"`
+}
+
+// eventMapping maps a connection to a Metricbeat event using connectionsSchema
+func eventMapping(content map[string]interface{}, fieldsSchema s.Schema) (mb.Event, error) {
+	fields, err := fieldsSchema.Apply(content)
+	if err != nil {
+		return mb.Event{}, errors.Wrap(err, "error applying connection schema")
+	}
+
+	err = util.UpdateDuration(fields, "uptime")
+	if err != nil {
+		return mb.Event{}, errors.Wrap(err, "failure updating uptime key")
+	}
+
+	err = util.UpdateDuration(fields, "idle_time")
+	if err != nil {
+		return mb.Event{}, errors.Wrap(err, "failure updating idle_time key")
+	}
+
+	moduleFields, err := moduleSchema.Apply(content)
+	if err != nil {
+		return mb.Event{}, errors.Wrap(err, "error applying module schema")
+	}
+
+	timestamp, err := util.GetNatsTimestamp(moduleFields)
+	moduleFields.Delete("server.time")
+	if err != nil {
+		return mb.Event{}, errors.Wrap(err, "failure parsing server timestamp")
+	}
+	event := mb.Event{
+		MetricSetFields: fields,
+		ModuleFields:    moduleFields,
+		Timestamp:       timestamp,
+	}
+	return event, nil
+}
+
+// eventsMapping maps per-connection metrics
+func eventsMapping(r mb.ReporterV2, content []byte) error {
+	var err error
+	connections := Connections{}
+	if err = json.Unmarshal(content, &connections); err != nil {
+		return errors.Wrap(err, "failure parsing NATS connections API response")
+	}
+
+	for _, con := range connections.Connections {
+		var evt mb.Event
+		con["server_id"] = connections.ServerID
+		con["now"] = connections.Now
+		evt, err = eventMapping(con, connectionsSchema)
+		if err != nil {
+			r.Error(errors.Wrap(err, "error mapping connection event"))
+			continue
+		}
```

Comment:
I think we could define `connections.Now` as `time.Time` so it is already parsed as time by `json.Unmarshal`. And then we could directly do here `evt.Timestamp = connections.Now`.

This way we wouldn't need to pass `con["now"]`, apply the schema, `util.GetNatsTimestamp()`, and delete `server.time`.

Reply:
Hmm, that's a good idea however is not directly applicable on all metricsets since most of them do not unmarshal on a struct but directly use the Schema on a `map[string]interface{}`. If we aim to apply this change too on the other metricsets maybe we need to get rid of the Schema 🤔 ?